src/github.com/matrix-org/dendrite/roomserver/storage/storage.go

```diff
@@ -35,3 +36,68 @@ func (d *Database) PartitionOffsets(topic string) ([]types.PartitionOffset, erro
 func (d *Database) SetPartitionOffset(topic string, partition int32, offset int64) error {
 	return d.statements.upsertPartitionOffset(topic, partition, offset)
 }
+
+// StoreEvent implements input.EventDatabase
+func (d *Database) StoreEvent(event gomatrixserverlib.Event) error {
+	var (
+		roomNID          int64
+		eventTypeNID     int64
+		eventStateKeyNID int64
+		eventNID         int64
+		err              error
+	)
+
+	if roomNID, err = d.assignRoomNID(event.RoomID()); err != nil {
+		return err
+	}
+
+	if eventTypeNID, err = d.assignEventTypeNID(event.Type()); err != nil {
+		return err
+	}
+
+	eventStateKey := event.StateKey()
+	if eventStateKey != nil {
+		if eventStateKeyNID, err = d.assignStateKeyNID(*eventStateKey); err != nil {
+			return err
+		}
+	}
+
+	eventID := event.EventID()
+	referenceSHA256 := event.EventReference().EventSHA256
+
+	if eventNID, err = d.statements.insertEvent(roomNID, eventTypeNID, eventStateKeyNID, eventID, referenceSHA256); err != nil {
+		return err
+	}
+
+	return d.statements.insertEventJSON(eventNID, event.JSON())
+}
+
+func (d *Database) assignRoomNID(roomID string) (roomNID int64, err error) {
+	if roomNID, err = d.statements.selectRoomNID(roomID); err != nil {
+		return
+	}
+	if roomNID == 0 {
+		return d.statements.insertRoomNID(roomID)
+	}
+	return
+}
+
+func (d *Database) assignEventTypeNID(eventType string) (eventTypeNID int64, err error) {
+	if eventTypeNID, err = d.statements.selectEventTypeNID(eventType); err != nil {
+		return
+	}
+	if eventTypeNID == 0 {
+		return d.statements.insertEventTypeNID(eventType)
```

Comment:
Boilerplate is boilerplatey, and unclear boilerplate at that.

I understand that we're exchanging "stuff" for NIDs, but I don't understand:
 - Which things can be exchanged for NIDs.
 - Why we want to do this.

As an aside, I dislike how we "mostly" use named-returns, except when we don't (this line). I'd rather we just used normal returns:
```go
whateverNID, err := d.statements.selectWhateverNID(whatever)
if err != nil {
  return 0, err
}
if whateverNID == 0 {
  return d.statements.insertWhateverNID(whatever)
}
return whateverNID, err
```

Also, the idea that "0" is a special sentinel value for NID feels like a weird choice, given it's more typical to return error consts like `sql.ErrNoRows`.